source/opt/if_conversion.cpp

```diff
@@ -0,0 +1,209 @@
+        context()->set_instr_block(select.get(), &block);
+        context()->ReplaceAllUsesWith(phi->result_id(), select->result_id());
+        iter.InsertBefore(std::move(select));
+        to_kill.push_back(phi);
+
+        return true;
+      });
+    }
+  }
+
+  for (auto inst : to_kill) {
+    context()->KillInst(inst);
+  }
+
+  return modified ? Status::SuccessWithChange : Status::SuccessWithoutChange;
+}
+
+bool IfConversion::CheckPhiUsers(ir::Instruction* phi, ir::BasicBlock* block) {
+  return get_def_use_mgr()->WhileEachUser(phi, [block,
+                                                this](ir::Instruction* user) {
+    if (user->opcode() == SpvOpPhi && context()->get_instr_block(user) == block)
+      return false;
+    return true;
+  });
+}
+
+uint32_t IfConversion::SplatCondition(analysis::Vector* vec_data_ty,
+                                      ir::BasicBlock* block,
+                                      ir::BasicBlock::iterator where,
+                                      uint32_t cond) {
+  // If the data inputs to OpSelect are vectors, the condition for
+  // OpSelect must be a boolean vector with the same number of
+  // components. So splat the condition for the branch into a vector
+  // type.
+  analysis::Bool bool_ty;
+  analysis::Vector bool_vec_ty(&bool_ty, vec_data_ty->element_count());
+  uint32_t bool_vec_id =
+      context()->get_type_mgr()->GetTypeInstruction(&bool_vec_ty);
+  std::vector<ir::Operand> ops;
+  for (size_t i = 0; i != vec_data_ty->element_count(); ++i) {
+    ops.emplace_back(SPV_OPERAND_TYPE_ID,
+                     std::initializer_list<uint32_t>{cond});
+  }
+  uint32_t splat_id = TakeNextId();
+  std::unique_ptr<ir::Instruction> splat(new ir::Instruction(
+      context(), SpvOpCompositeConstruct, bool_vec_id, splat_id, ops));
+  context()->set_instr_block(splat.get(), block);
+  get_def_use_mgr()->AnalyzeInstDefUse(splat.get());
+  where.InsertBefore(std::move(splat));
+  return splat_id;
+}
+
+bool IfConversion::CheckType(uint32_t id) {
+  ir::Instruction* type = get_def_use_mgr()->GetDef(id);
+  SpvOp op = type->opcode();
+  if (spvOpcodeIsScalarType(op) || op == SpvOpTypePointer ||
+      op == SpvOpTypeVector)
+    return true;
+  return false;
+}
+
+ir::BasicBlock* IfConversion::GetBlock(uint32_t id) {
+  return context()->get_instr_block(get_def_use_mgr()->GetDef(id));
+}
+
+ir::BasicBlock* IfConversion::GetIncomingBlock(ir::Instruction* phi,
+                                               uint32_t predecessor) {
+  uint32_t in_index = 2 * predecessor + 1;
+  return GetBlock(phi->GetSingleWordInOperand(in_index));
+}
+
+ir::Instruction* IfConversion::GetIncomingValue(ir::Instruction* phi,
+                                                uint32_t predecessor) {
+  uint32_t in_index = 2 * predecessor;
+  return get_def_use_mgr()->GetDef(phi->GetSingleWordInOperand(in_index));
+}
+
+ir::BasicBlock* IfConversion::CommonDominator(
```

Comment:
Would this be better in the Dominator Tree class?  I could see many other things using this.